### PR TITLE
Fix character values - do not allow values less than 1 - fixes #8512

### LINKED ISSUE #8512: Fix Character Values allows values less than 1
[//]: # (Before logging this issue, look through common problems at https://github.com/HabitRPG/habitrpg/issues If you find your issue there, read at least the first post to see if there is a workaround for you)

[//]: # (Github is primarily used for reporting bugs. If you have a feature request, use "Help > Request a Feature" so that the feature request can be vetted by the larger Habitica community)

[//]: # (To report a bug in one of the mobile apps, please report it in the correct repository. Android: https://github.com/HabitRPG/habitrpg-android, iOS: https://github.com/HabitRPG/habitrpg-ios)

[//]: # (For more guidelines see https://github.com/HabitRPG/habitrpg/issues/2760)

[//]: # (Fill out relevant information - UUID is found in Settings -> API)
General Info
  * UUID: e8277c2b-d1a0-47f8-9a08-31211ee578c8
  * Browser: Chrome
  * OS: Windows 10

### Description
[//]: # (Describe bug in detail here. Include pictures if helpful.)

The "Fix Character Values" feature on the Site page disallows decimal values for Level, which is correct, but it still allows you to fill in 0 or a negative number. Levels below 1 are not supported and could cause unexpected behavior with other features related to game statistics.

image

#### Console Errors
[//]: # (Include any JavaScript console errors here.)

None

### PR DESCRIPTION
**Pull request by cailinglu**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8512 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
- ~~Display an error message if invalid character values are entered in the 'fix character values' section~~
- ~~I put an invalidValue message in the en locale settings.json file that says: "Invalid value: Level must be 1 or greater. All other values must be 0 or greater."~~
  - ~~Let me know if this should be modified. This was what I have as a general message to display, rather than displaying a separate message depending on which values are invalid.~~
- Display an error message if a level less than 1 is entered in the 'fix character values' section
- Fix character values modal only closes when the user clicks on 'save and close' and the entered values are valid (can still be closed if they hit the 'cancel' button)

levelerror

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: a3f88ebe-efd8-4ff9-a887-efb5d2522a97

### CODE REVIEW

**Comment by Alys** [comment]
@caillu This looks like a good fix for the level issue, and I like that you've expanded the fix to other values as well. However Experience is actually allowed to be negative in normal game play and so it should be allowed to be negative in Fix Character Values (FCV) too (reasons below). Health can be negative briefly in normal game play, but the case for allowing that in FCV is weaker.

Experience can become negative if you level up from completing a task and then realise that you'd accidentally marked the wrong task as complete. When you uncomplete the task, the XP gained is subtracted, sometimes leading to a negative value. This doesn't do any harm because the XP increases normally as you do other tasks and rises above zero once you've earned back enough of it, and [we made a decision](https://github.com/HabitRPG/habitica/issues/5736#issuecomment-127677025) to allow negative XP in that situation because it was the clearest, simplest behaviour when unticking a task that had caused a level-up. Because negative XP can be obtained from normal usage of Habitica, FCV should allow it. I can think of a couple of uses for setting XP negative in FCV: a Socialite or wiki editor might want to set their XP negative temporarily to obtain a screenshot for illustrating that situation on the wiki. Or a player who had obtained negative XP through normal gameplay might then have had their XP increased dramatically due to overenthusiastic buffing from a party member, and might want to set the XP back to the real value so as to avoid feeling like they were cheating. Also, players who are using FCV to make the game harder for themselves might want to set their XP negative each time they level up - "hard mode" tweaks like that are not uncommon (we actually have at least one guild dedicated to them).

Health can be negative if you take enough damage from missed Dailies or a quest boss. In that situation, it becomes 50 again as soon as you take any action on the site (you die and then revive immediately). The only time you ever see negative health is very briefly on the Experience bar and then in the death modal. I imagine that a wiki editor _might_ possibly want to get a screenshot of that, but really a screenshot with a health of exactly zero would be good enough, so if we want to prohibit negative values of Health in FCV, I'm okay with that.

I think it's okay to have a single message for all error conditions in FCV, although it will get a little wordy since XP can be negative. I guess something like "Invalid value: Level must be 1 or greater. Experience may be negative or 0 or greater. All other values must be 0 or greater."

**Comment by cailinglu** [comment]
I went back and made changes. Decided to only check if the level is valid since that was the original issue, and if other values being negative aren't impacting the game, then perhaps I should just leave them as is for now.

File: website/client-old/js/controllers/settingsCtrl.js [hunk 1]
```diff
@@ -164,14 +164,17 @@ habitrpg.controller('SettingsCtrl',
...
         achievements = $scope.restoreValues.achievements;
-      User.set({
-        "stats.hp": stats.hp,
-        "stats.exp": stats.exp,
-        "stats.gp": stats.gp,
-        "stats.lvl": stats.lvl,
-        "stats.mp": stats.mp,
-        "achievements.streak": achievements.streak
-      });
+      if (_validateValues(stats.lvl)) {
```
Comment (crookedneighbor) [inline website/client-old/js/controllers/settingsCtrl.js]: 1) this method is called `_validateValues` but only validates a single value, level.

2) Wouldn't this be cleaner:

```
if (stats.lvl < 1) {
  Notification.error(env.t('invalidLevel'), true);
  return;
}

User.set({ /* stuff */ });
```

**Comment by crookedneighbor** [comment]
TBH, we should be validating this on the server if we want to prevent invalid values: https://github.com/HabitRPG/habitica/blob/164177f010a9791046d300f5053ddae664bcf251/website/server/models/user/schema.js#L493-L497

**Comment by cailinglu** [comment]
Should I also be editing code on the server side then?

**Comment by crookedneighbor** [comment]
If the problem you're trying to solve is getting the user into a bad state, then yes, I think we should be validating on the server since a user could use the API to make this same call and put themselves in a bad state.

**Comment by crookedneighbor** [comment]
`getting the user into a bad state` -> `preventing the user from getting into a bad state`

**Comment by cailinglu** [comment]
I looked at the schema linked and just added a `min: 1` for the user's level. Not 100% sure if I should be doing more.

**Comment by TheHollidayInn** [comment]
Looks good to me

File: website/server/models/user/schema.js [hunk 2]
```diff
@@ -494,7 +494,7 @@ let schema = new Schema({
     mp: {type: Number, default: 10},
     exp: {type: Number, default: 0},
     gp: {type: Number, default: 0},
-    lvl: {type: Number, default: 1},
+    lvl: {type: Number, default: 1, min: 1},
```
Comment (paglias) [inline website/server/models/user/schema.js:497]: should we run a migration for users who have it set to a value < 1?

Comment (Alys) [reply website/server/models/user/schema.js:497]: No need in my opinion. The site behaves normally with negative levels as far as I can tell from a quick test, and even if there were problems, there'd be no need for a migration before merging this (and we'd probably already have heard from any users affected by it).

**Comment by paglias** [comment]
Well with this change users with a level < 1 won't be able to use the site,
saving the user will fail. Hmm.. I think it depends on whether the
validation is done always or only when the field changes

Il 09 Apr 2017 21:20, "Alys" <notifications@github.com> ha scritto:

**Comment by Alys** [comment]
Fair point. I'll do a search for users with level < 1 now and another after we merge. I'd expect the numbers to be so low it's not worth writing a dedicated migration script.

**Comment by Alys** [comment]
I've tested locally as listed below and it's good.  I have not reviewed the code.

1. Used FCV to set level to 0 - failed: level was not changed and message appeared "Invalid value: Level must be 1 or greater."
1. Used FCV to set level to -1 - failed: as above
1. Used FCV to set level to -10 - failed: as above
1. Used FCV to set level to 1 - succeeded: level was set correctly
1. Used API (example call below) to set level to -1 - failed: level was not changed and the error shown below was given
1. Used API to set level to 0 - failed: as above
1. Used API to set level to 1 - succeeded: level was set correctly

curl -X PUT -H "x-api-user: d2c49eab-dae4-4ffb-9d40-cccb4d793cdf" -H "x-api-key: b6b6b446-d5eb-468e-a467-d640118c9661" http://localhost:3001/api/v3/user -d "stats.lvl=-1"  

{"success":false,"error":"BadRequest","message":"User validation failed","errors":[{"message":"Path `stats.lvl` (-1) is less than minimum allowed value (1).","path":"stats.lvl","value":-1}]}

File: test/client-old/spec/controllers/settingsCtrlSpec.js [hunk 3]
```diff
@@ -334,4 +336,42 @@ describe('Settings Controller', function () {
...
+            exp: 0,
+            gp: 0,
+            lvl: 0,
+            mp: 0,
+          },
+          achievements: {
+            streak: 0,
+          },
+        };
+        scope.restore();
```
Comment (TheHollidayInn) [inline test/client-old/spec/controllers/settingsCtrlSpec.js]: I'm not sure this is correct. You have stubbed restore above, so the function won't be called, thus, you expectation below will never be true leading to this test not doing anything.

**Comment by TheHollidayInn** [comment]
Noted towards you tier! Thanks, @caillu !

**Pull request merged**